### ISSUE TITLE
"Database is locked" error when removing agents

### ISSUE THREAD

**Issue by mgmacias95** [wazuh/wazuh-api#311]
Hello team,

The following error is returned when trying to remove agents in a big environment:
```javascript
    {
            "id": "002",
            "error": {
               "message": "Maximum attempts exceeded for sqlite3 execute: database is locked",
               "code": 2002
            }
         },
         {
            "id": "003",
            "error": {
               "message": "Maximum attempts exceeded for sqlite3 execute: database is locked",
               "code": 2002
            }
         },
         {
            "id": "004",
            "error": {
               "message": "Maximum attempts exceeded for sqlite3 execute: database is locked",
               "code": 2002
            }
         },
         {
            "id": "005",
            "error": {
               "message": "Maximum attempts exceeded for sqlite3 execute: database is locked",
               "code": 2002
            }
         }
      ],
      "total_failed_ids": 128,
      "msg": "Some agents were not removed",
      "older_than": "7d"
   }
```

I think we should prevent this error or, at least, return a more user friendly error description.

Best regards,
Marta

**Comment by pm17788** [comment]
I've seen this happen in my cluster (my management cluster members are on VMs), and I've also seen the same error show up when doing API queries for things `GET /syscheck/:agent_id/last_scan`.

In `/var/ossec/logs/ossec.log`, it shows up like this:
```
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10195' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10196' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10197' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10198' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10199' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10200' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10201' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10202' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10203' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10204' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10205' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10206' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10207' group.             
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10208' group.                                                                                                                                                  
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10209' group.                                                          
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10210' group.                                                          
2019/02/08 21:04:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '10211' group.             
```

If I crank up the verbosity/debug levels, I'll see the timeouts.

**Comment by mgmacias95** [comment]
Hello @pm17788,

After talking to the @wazuh/core team, it seems this issue happens when removing an agent:
1. The agent is first removed from the [client.keys](https://documentation.wazuh.com/current/development/client-keys.html) file.
2. All agent files used to store information such as group assignment, OS, last keep alive... are removed afterwards.

It seems this message is shown at some moment between steps 1 and 2. We will do further research to prevent this in the future. Thank you so much for reporting!

Best regards,
Marta

**Comment by pm17788** [comment]
Hiya @mgmacias95,

Thank you for the follow up. In my circumstance, I see the messages show up on *some* (but not all) members of the 6-worker/1-master cluster, which doesn't seem to line up with the theory that the messages are indicative of a race condition between the removal of an agent from `client.keys` and the underlying SQLite3 databases being cleaned up.

For example, looking at the logs for worker5, I see:
```2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12600' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12601' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12602' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12603' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12604' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12605' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12606' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12607' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12608' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12609' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12610' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12611' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12612' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12613' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12614' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12615' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12616' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12617' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12618' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12619' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12620' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12621' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12622' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12623' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12624' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12626' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12627' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12628' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12629' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12630' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12631' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12632' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12633' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12634' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12635' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12636' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12637' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12638' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12640' group.
2019/02/14 21:48:43 wazuh-modulesd:database: ERROR: Couldn't sync agent '12641' group.
2019/02/14 21:52:03 wazuh-modulesd:database: ERROR: Couldn't sync agent '12642' group.
```

However, all of these agent IDs are active in my environment (notice how the sample agent is on node `worker7`, though the error was logged on `worker5`? Not sure if this correlative in any way)

```
root@dlt-ossec-1:/var/ossec/logs# /var/ossec/bin/cluster_control -a | grep 12621
  ID: 12621, Name: server3805, IP: any, Status: Active,  Node: worker7
```

Cheers,

Pavel

**Comment by mgmacias95** [comment]
Hello,

I am working on a fix for this issue on https://github.com/wazuh/wazuh/pull/2869. The proposed fix (https://github.com/wazuh/wazuh/pull/2869/commits/e5bf786b4499355f1c34b9b8e1f52796a551070c) consists on adding a sleep interval everytime a `database locked` exception is raised. I have tested this approach and it's more likely to get a successful result when the databse is locked.

Best regards,
Marta

**Comment by crd1985** [comment]
Close this issue since it should be fixed in wazuh/wazuh#2869
